client side ALPN for OSX

wfurt opened the pull request:
This change implements client side ALPN for OSX. 
Server side is currently not supported by coretls on OSX. (see related issue) 

This uses functions available via private header. With 10.13.4, one could use new API but that fails to link on all previous versions. With using this, we can get consistent behavior on all supported OSX versions. 

related to #30492

wfurt committed:
ALPN client side for OSX

wfurt commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -388,6 +390,12 @@ internal enum TlsCipherSuite
             out int osStatus);
 
+        [DllImport(Interop.Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_SslCtxSetAlpnProtos")]
+        internal static extern int SslCtxSetAlpnProtos(SafeSslHandle ctx, IntPtr protos, int len, out int osStatus);
```
Instead of IntPtr and `fixed` this can be `ref byte` and you can call it with `ref buffer[0]` (provided buffer.Length isn't 0)

wfurt replied:
This is what we do for OpenSSL. (src/Interop/Unix/System.Security.Cryptography.Native/Interop.SslCtx.cs) 
Should I change that as well? It seems reasonable to keep things consistent.
(This also applied to some comments bellow)

bartonjs replied:
If it was just changing the P/Invoke target then my normal tidiness heuristics would say that's a different cleanup change.

But since it looks like the ConvertAlpnProtocolListToByteArray method was a copy/paste, it should probably be factored out into shared source (vs copy/paste), and since you're now editing Interop.SslCtx.cs anyways you may as well change the P/Invoke :smile:.

jkotas replied:
> this can be `ref byte`

We should be using `ref byte` for unsafe buffers. It is not capturing the intent. It is more appropriate to use use `byte*`.

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -569,6 +577,60 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
+            foreach (SslApplicationProtocol protocol in applicationProtocols)
+            {
+                buffer[offset++] = (byte)(protocol.Protocol.Length);
+                protocol.Protocol.Span.CopyTo(new Span<byte>(buffer).Slice(offset));
```
`buffer.AsSpan(offset)` is a bit more compact than `new Span<byte>(buffer).Slice(offset)`

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -569,6 +577,60 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
+            }
+
+            byte[] buffer = new byte[protocolSize];
+            var offset = 0;
```
This is an inappropriate `var`.  I'm pretty sure that it's an `int`, but our style rules only allow var when casting or using operator-new.

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -569,6 +577,60 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
+            {
+                buffer[offset++] = (byte)(protocol.Protocol.Length);
+                protocol.Protocol.Span.CopyTo(new Span<byte>(buffer).Slice(offset));
+                offset += protocol.Protocol.Length;
```
Rather than continually switching to a Span it seems like what you want is something like

```C#
byte[] buffer = new byte[protocolSize];
Span<byte> dest = buffer;

foreach (SslApplicationProtocol protocol in applicationProtocols)
{
    dest[0] = (byte)protocol.Protocol.Length;
    dest = dest.Slice(1);

    protocol.Protocol.Span.CopyTo(dest);
    dest = dest.Slice(protocol.Protocol.Length);
}

return buffer;
```

Alternatively, keeping the code mostly as it is (with a tracked offset) but saving the Span outside the loop.

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -4,6 +4,16 @@
 #include "pal_ssl.h"
 
+// This would come from private header  <Security/SecureTransportPriv.h>
+// 10.13.4 introduced public API but linking would fail on all prior versions.
```
We don't call private API.

If this is lightup, though, that's fine.  (And if the lightup happens to unexpectedly work on earlier versions, that's fine, too).

If it's public API in newer versions than our build machine, that's just lightup.  (And the comment should be adjusted accordingly)

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -4,6 +4,16 @@
+// This can be revisited after we drop support for 10.12.
+// Also note, this works only on client side in coretls-155 and lower.
+
+const void * SSLGetALPNData(SSLContextRef contex, size_t *length);
```
Is this declaration going to cause a problem when we build on a Mojave machine?

bartonjs replied:
I feel like `clang-format` would have written this as `void*`, not `void *`.  I *think* this file is already otherwise clang-format formatted.

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -4,6 +4,16 @@
+const void * SSLGetALPNData(SSLContextRef contex, size_t *length);
+OSStatus SSLSetALPNData(SSLContextRef context, const void *data, size_t length);
+
+// end of private ALPN.
```
ALPN lightup

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -153,6 +163,42 @@ int32_t AppleCryptoNative_SslSetTargetName(SSLContextRef sslContext,
+
+    // Get protocol data from TLS extension
+    *len = (uint32_t)alpnLen - 1;
+    *protocol = ext + 1;
```
Why the `-1`/`+1`?

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -153,6 +163,42 @@ int32_t AppleCryptoNative_SslSetTargetName(SSLContextRef sslContext,
+    return *pOSStatus == noErr;
+}
+
+DLLEXPORT int32_t AppleCryptoNative_SslGetAlpnSelected(SSLContextRef sslContext, const uint8_t** protocol, uint32_t* len)
```
Set takes a signed int, get outs an unsigned int.  Why the discrepancy?

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -153,6 +163,42 @@ int32_t AppleCryptoNative_SslSetTargetName(SSLContextRef sslContext,
+    }
+
+    // Get protocol data from TLS extension
+    *len = (uint32_t)alpnLen - 1;
```
Missing an overflow check.  `SizeTToInt32` would do it if the output was a signed integer.

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.h:
```diff
@@ -138,6 +138,24 @@ DLLEXPORT int32_t AppleCryptoNative_SslSetTargetName(SSLContextRef sslContext,
+DLLEXPORT int32_t AppleCryptoNative_SslCtxSetAlpnProtos(SSLContextRef sslContext,
+                                                     const uint8_t* protocols,
+                                                     int32_t len,
+                                                     int32_t* pOSStatus);
```
Clang-format, please.

bartonjs commented on src/System.Net.Security/tests/FunctionalTests/SslStreamAlpnTests.cs:
```diff
@@ -180,6 +189,38 @@ public async Task SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail()
         }
 
+        [OuterLoop("Uses external server")]
+        [PlatformSpecific(TestPlatforms.OSX)]
```
This test doesn't seem to be macOS specific, why not have it run for all platforms?

bartonjs commented on src/System.Net.Security/tests/FunctionalTests/SslStreamAlpnTests.cs:
```diff
@@ -180,6 +189,38 @@ public async Task SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail()
+                        SslClientAuthenticationOptions clientOptions = new SslClientAuthenticationOptions
+                        {
+                            ApplicationProtocols = new List<SslApplicationProtocol> { SslApplicationProtocol.Http2 ,  SslApplicationProtocol.Http11 },
+                            TargetHost =  server.Host
```
Some interesting interior whitespace in these lines.

bartonjs commented on src/System.Net.Security/tests/FunctionalTests/System.Net.Security.Tests.csproj:
```diff
@@ -47,6 +47,9 @@
     <Compile Include="$(CommonTestPath)\System\Net\Configuration.Certificates.cs">
       <Link>Common\System\Net\Configuration.Certificates.cs</Link>
+    </Compile>
+        <Compile Include="$(CommonTestPath)\System\Net\Configuration.Http.cs">
```
Please fix the indentation here.

krwq commented on src/System.Net.Security/tests/FunctionalTests/SslStreamAlpnTests.cs:
```diff
@@ -180,6 +189,38 @@ public async Task SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail()
+                catch (Exception e)
+                {
+                    // Failures to connect do not cause test failure.
+                    _output.WriteLine("Unable to connect: {0}", e);
```
I think writing to stdout works correctly (at least worked for me last time on windows and linux), having the exception directly tells you the exact place and makes it slightly easier to debug (not sure if helps in async scenario as in here)

wfurt replied:
Writing to xunit handle allows to preserve it in test result database. (as it gets written to result xml and processed by CI)
I used stdout only for local debug and there are already other tests using this pattern.

davidsh replied:
Yes, we should not be using stdout in Xunit.  We should only be using the `ITestOutputHelper` pattern.

wfurt committed:
rework code to use new API introduced in 10.13.4 + feedback from review

wfurt commented:
I updated code to use new API and this will silently fail on OSX versions prior 10.13.4 (so as it does today).
Working with CoreFoundation seems expensive so I added code to pre-create objects for common cases to avid allocations for each ssl session.

wfurt commented:
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop OSX x64 Debug Build

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -579,56 +590,76 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
+                {
+                    // we did not match common case. This is more expensive path allocating Core Foundation objects.
+                    cfProtocolRefs = new  SafeCreateHandle[protocols.Count];
+                    IntPtr[] protocolsPtr = new  System.IntPtr[protocols.Count];
```
NIt: two spaces after the `new` operator (thought I was having a font or eyeball problem)

bartonjs commented on src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs:
```diff
@@ -579,56 +590,76 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
-
-            fixed (byte* b = buffer)
+            SafeCreateHandle cfProtocolsRefs = null;
+            SafeCreateHandle[] cfProtocolRefs = null;
```
Maybe the first one should be `cfProtocolsRef`, or `cfProtocolsArrayRef`.  I was not noticing the interior `s` difference in the `else` case, and was greatly confused.

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_ssl.c:
```diff
@@ -163,40 +161,46 @@ int32_t AppleCryptoNative_SslSetTargetName(SSLContextRef sslContext,
-    return 1;
+    if (osStatus == noErr && protocols != NULL && CFArrayGetCount(protocols) > 0)
+    {
+        *protocol = CFStringCreateExternalRepresentation(NULL, CFArrayGetValueAtIndex(protocols, 0), kCFStringEncodingASCII, 0);
```
If you haven't already, please run `clang-format -i` on this file.

bartonjs commented on src/System.Net.Security/tests/FunctionalTests/SslStreamAlpnTests.cs:
```diff
@@ -190,8 +191,7 @@ public async Task SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail()
-        [Theory]
+        [ConditionalTheory(nameof(ClientSupportsAlpn))]
         [MemberData(nameof(Http2Servers))]
         public async Task SslStream_OSX_Alpn_Success(Uri server)
```
Now that the test isn't platform specific, the name should probably be changed.  s/OSX/Client/, perhaps?

wfurt committed:
reviews feedback round 2

wfurt commented:
@dotnet-bot test Outerloop OSX x64 Debug Build

wfurt committed:
feedback from review
